pandas/core/reshape/pivot.py

```diff
@@ -148,7 +148,8 @@ def pivot_table(
         table = table.sort_index(axis=1)
 
     if fill_value is not None:
-        table = table.fillna(value=fill_value, downcast="infer")
+        filled = table.fillna(value=fill_value, downcast="infer")
+        table = cast(DataFrame, filled)
```

Comment:
I think the preference is for assert instead of cast @jbrockmendel 
```suggestion
        assert table is not None  # needed for mypy
```

Reply:
The error is `` error: Incompatible types in assignment (expression has type "Optional[DataFrame]", variable has type "DataFrame")``. I cant reassign ``table`` to a broader type. So that pattern won't work.

Could do

```
filled = table.fillna(value=fill_value, downcast="infer")
assert filled is not None
table = filled
```
 but that's also ugly. But I agree that asserts are better, so I'll do that (unless you got something less ugly :-))

Reply:
> (unless you got something less ugly :-)

we probably should be overloading all methods that have an inplace parameter, but that would require the use of Literal.

Reply:
Literals won’t come until 3.8 is the minimum version, so will be a while. Should we allow assignment to incompatible type in mypy, until 3.8? Seems like a lesser evil?

Reply:
Ok, should probably use ignore commands in the first line: https://mypy.readthedocs.io/en/latest/error_codes.html#silencing-errors-based-on-error-codes. Then table =... can be used and the third line can be dropped.

I’ll update later to use this pattern.

Reply:
we are using the default for allow_redefinition. It does result in lots of variable renaming and some messy code, but the additional strictness is probably worth it.

Reply:
Is there any way we could get mypy to understand that without `inplace=True` `isna` returns non-None?

Reply:
Not until 3.8 and Literal, AFAIK.

There has been an idea floating around that the ``inplace`` parameter should be deprecated everywhere (unless there is a performance benefit to it). That would simplify the return types.

Reply:
I've changed this section to use ``# type: ignore``  directive.

Reply:
it's a good idea, we need to move this forward. see #29197 could add # noqa for now.